### PR TITLE
Reimplement UpdateFlags as enum-like class

### PR DESCRIPTION
**Pull request by masterleinad**
This is based on the approach described in https://github.com/dealii/dealii/pull/13229#issuecomment-1015531116 to provide an enum-like class for `UpdateFlags` in a (mostly) backward-compatible way.

For now, this class also defines a contains member function but that is, of course, up for discussion and it shouldn't be difficult to change it into an `operator bool()` operator.

Based on and supersedes #13229.

### CODE REVIEW

File: .clang-tidy [hunk 1]
```diff
@@ -8,6 +8,6 @@
...
 # - performance-no-automatic-move: All modern compiler perform the return value optimization and we prefer to keep things const.
 
-Checks: "-*,cppcoreguidelines-pro-type-static-cast-downcast,google-readability-casting,modernize-*,-modernize-pass-by-value,-modernize-raw-string-literal,-modernize-use-auto,-modernize-use-override,-modernize-use-default-member-init,-modernize-use-transparent-functors,-modernize-use-trailing-return-type,-modernize-use-nodiscard,-modernize-avoid-c-arrays,-modernize-concat-nested-namespaces,use-emplace,mpi-*,performance-*,-performance-inefficient-string-concatenation,-performance-no-automatic-move"
+Checks: "-*,cppcoreguidelines-pro-type-static-cast-downcast,google-readability-casting,modernize-*,-modernize-pass-by-value,-modernize-raw-string-literal,-modernize-use-auto,-modernize-use-override,-modernize-use-default-member-init,-modernize-use-transparent-functors,-modernize-use-trailing-return-type,-modernize-use-nodiscard,-modernize-avoid-c-arrays,-modernize-concat-nested-namespaces,use-emplace,mpi-*,performance-*,-performance-inefficient-string-concatenation,-performance-no-automatic-move,readability-implicit-bool-conversion"
```
Comment (bangerth) [inline .clang-tidy:11]: I think we should be able to re-enable this independent of the current patch, right?

File: include/deal.II/fe/fe_poly_tensor.h [hunk 2]
```diff
@@ -301,8 +301,8 @@ class FE_PolyTensor : public FiniteElement<dim, spacedim>
     std::vector<Tensor<4, dim>> third_derivatives(0);
     std::vector<Tensor<5, dim>> fourth_derivatives(0);
 
-    if ((update_flags & (update_values | update_gradients | update_hessians)) !=
-        0u)
+    if (update_flags.contains(
+          (update_values | update_gradients | update_hessians)))
```
Comment (bangerth) [inline include/deal.II/fe/fe_poly_tensor.h:305]: You can remove one level of parentheses here.

File: include/deal.II/fe/fe_update_flags.h [hunk 3]
```diff
@@ -33,6 +33,13 @@ DEAL_II_NAMESPACE_OPEN
 #ifndef DOXYGEN
 template <int, int>
 class FiniteElement;
+class UpdateFlags;
+
+namespace internal
+{
+  constexpr UpdateFlags
+  make_update_flags(int);
```
Comment (bangerth) [inline include/deal.II/fe/fe_update_flags.h:41]:
```suggestion
  make_update_flags(const int);
```

File: include/deal.II/fe/fe_update_flags.h [hunk 4]
```diff
@@ -119,114 +176,119 @@ enum UpdateFlags
...
+   * Private constructor only to be used internally.
+   */
+  constexpr explicit UpdateFlags(int i);
+
+  /**
+   * Friend declaration so that make_update_flags can call the private
+   * constructor.
+   */
+  friend constexpr UpdateFlags
+  internal::make_update_flags(int);
```
Comment (bangerth) [inline include/deal.II/fe/fe_update_flags.h:291]:
```suggestion
  internal::make_update_flags(const int);
```

File: include/deal.II/fe/fe_update_flags.h [hunk 5]
```diff
@@ -240,114 +302,143 @@ inline StreamType &
 operator<<(StreamType &s, const UpdateFlags u)
 {
   s << " UpdateFlags|";
-  if (u & update_values)
+  if (u & UpdateFlags::update_values)
```
Comment (bangerth) [inline include/deal.II/fe/fe_update_flags.h:305]: Shouldn't this use `contains` as well? Out of curiosity, why do you need the classname qualification?

File: include/deal.II/fe/fe_update_flags.h [hunk 6]
```diff
@@ -240,114 +302,143 @@ inline StreamType &
...
     s << "jacobian_pushed_forward_3rd_derivatives|";
 
   // TODO: check that 'u' really only has the flags set that are handled above
   return s;
 }
 
+namespace internal
+{
+  constexpr UpdateFlags
+  make_update_flags(int field)
```
Comment (bangerth) [inline include/deal.II/fe/fe_update_flags.h:351]:
```suggestion
  make_update_flags(const int field)
```

File: include/deal.II/fe/fe_update_flags.h [hunk 7]
```diff
@@ -240,114 +302,143 @@ inline StreamType &
...
+
+inline constexpr bool
+UpdateFlags::contains(const UpdateFlags in) const
+{
+  return (field & in.field) != 0;
+}
+
+// Import all flags into the global namespace.
+constexpr UpdateFlags update_default;
+constexpr UpdateFlags update_values;
```
Comment (bangerth) [inline include/deal.II/fe/fe_update_flags.h:418]: Don't these have to be initialized somehow? I don't see where the "import" actually happens :-)

**Review by bangerth — COMMENTED** [review]
I haven't read through it all, but I kind of like it :-)

File: source/fe/fe_update_flags.cc [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+  update_quadrature_points | update_JxW_values | update_jacobians |
+  update_jacobian_grads | update_jacobian_pushed_forward_grads |
+  update_jacobian_2nd_derivatives |
+  update_jacobian_pushed_forward_2nd_derivatives |
+  update_jacobian_3rd_derivatives |
+  update_jacobian_pushed_forward_3rd_derivatives | update_inverse_jacobians |
+  update_boundary_forms | update_normal_vectors |
+  update_covariant_transformation | update_contravariant_transformation |
+  update_transformation_values | update_transformation_gradients |
+  update_volume_elements;
```
Comment (nfehn) [inline source/fe/fe_update_flags.cc:138]: What is the difference between this code and lines 21-78 above?

**Comment by drwells** [comment]
Given #13284 and the proliferation of update-flags like things in the library now I think we should go ahead and do something like this for improved type safety.

File: source/fe/fe_update_flags.cc (hunk 8 above)
Comment (masterleinad) [reply source/fe/fe_update_flags.cc]: It should have been the definition of the global variables. Fixed now.

File: include/deal.II/fe/fe_update_flags.h (hunk 5 above)
Comment (masterleinad) [reply include/deal.II/fe/fe_update_flags.h]: Yes, fixed. I used the class name due to some problems with the order of declarations/definitions. Using the class static variables worked better.

**Comment by nfehn** [comment]
I am very happy to see this proposal! I think this is a clear improvement. Let me share some further ideas below:

Since `UpdateFlags` contains a static member variable for every type of flag, this is essentially a design towards a list of boolean parameters from a design perspective (which was heavily criticized by @luca-heltai in https://github.com/dealii/dealii/issues/9888).

If I see it correctly, the reason why one does not use boolean parameters is backwards compatibility in terms of the bitwise enums/operations. The following aspects might be considered a disadvantage of the proposed design:

- A private constructor and a function declared "friend" is needed
- For a variable `UpdateFlags flags;`, one could write `flags.update_values.update_gradients` (potentially continuing this chain further), since sub-elements are again of the same type `UpdateFlags` (flags are nested), or do I miss something?

Forgetting about backward compatibility for a moment, this would suggest a different design where the inner type should be different from the outer type, which could be realized more easily by a list of boolean parameters for every type of flag. In that case, also the function `contains()` would no longer be necessary. Instead of

```
// nested approach: use flags
if (flags.contains( (update_values | update_gradients | update_hessians)))
{
  ...
}
```

one would write

```
// boolean approach: use flags
if (flags.update_values or flags.update_gradients or flags.update_hessians)
{
  ...
}
```

Another difference of both approaches is construction
```
// nested approach: construction
UpdateFlags flags = update_values | update_gradients | update_hessians;
FEValues fe_values(mapping, fe, quadrature, flags);
```
and

```
// boolean approach: construction
UpdateFlags flags; 
flags.update_values = true; 
flags.update_gradients = true; 
flags.update_hessians = true;
FEValues fe_values(mapping, fe, quadrature, flags);
```
Again, this was discussed controversially in https://github.com/dealii/dealii/issues/9888, where the first variant received a lot of applause. My personal point of view is that the second variant is not disadvantageous compared to the first one.

File: source/fe/fe_update_flags.cc (hunk 8 above)
Comment (nfehn) [reply source/fe/fe_update_flags.cc]: I think I still see the code twice, or which commit is it exactly?

**Comment by bangerth** [comment]
@nfehn Incompatibility is not an option here :-) Update flags are so widely used that it would amount to a herculean effort to change things. For example, ASPECT has 686 places where update flags are used. If you're only spending half a minute on each, it would still take 5 hours. And you'd then also be locked into a new deal.II version. So that's not going to happen :-)

**Comment by bangerth** [comment]
@masterleinad One of the comments by @nfehn made me think of the following ambiguity: If you write
```
  flags.contains (update_values | update_gradients)
```
does that mean that `flags` contains at least one, or all of these flags? My preference would be if `UpdateFlags::contains()` checks that only a single bit is set in the argument. If that's not the case, we error out and refer the user to two additional functions: `UpdateFlags::contains_any()` and `UpdateFlags::contains_all()`.

**Comment by nfehn** [comment]
> @masterleinad One of the comments by @nfehn made me think of the following ambiguity: If you write
> 
> ```
>   flags.contains (update_values | update_gradients)
> ```
> 
> does that mean that `flags` contains at least one, or all of these flags? My preference would be if `UpdateFlags::contains()` checks that only a single bit is set in the argument. If that's not the case, we error out and refer the user to two additional functions: `UpdateFlags::contains_any()` and `UpdateFlags::contains_all()`.

@bangerth Do you mean I should have written `and` instead of `or` in my example above? If so, this would indeed be a deficiency of the new design. Going for two different functions shows that the design is not intuitive enough in my opinion.

**Comment by masterleinad** [comment]
> @masterleinad One of the comments by @nfehn made me think of the following ambiguity: If you write
  flags.contains (update_values | update_gradients)
does that mean that flags contains at least one, or all of these flags?

We can't easily error out for this case since the return value of `operator |` is again `UpdateFlags` and I don't think we can change that either. In my opinion, the syntax in this particular case also makes sense since I would read
```
flags.contains (update_values | update_gradients)
```
as "flags contains update_values or update_gradients", i.e. at least one of these flags. (Of course, this interpretation doesn't work with `operator&`).

**Comment by nfehn** [comment]
> @nfehn Incompatibility is not an option here :-) Update flags are so widely used that it would amount to a herculean effort to change things. For example, ASPECT has 686 places where update flags are used. If you're only spending half a minute on each, it would still take 5 hours. And you'd then also be locked into a new deal.II version. So that's not going to happen :-)

I will not put as much energy into this as in previous discussions ... just one more comment: It would be great to discuss the topic first before excluding certain suggestions (I will make one more suggestion in the related issue ...). Tets were also claimed to never find their place in dealii ... Also, I am tempted to ask why ASPECT has 686 occurrences of UpdateFlags? Since you count hours: I do not want to count the hours that dealii developers invested into this topic in related issues and PRs up to now, discussions where I have the impression that they are governed by a fear of breaking compatibility instead of developing really good solutions.

**Comment by masterleinad** [comment]
I am fine with creating a new class that could replace `UpdateFlags` eventually if we find good reasons for it (I haven't read through all of the discussion in #9888).

AFAIK `UpdateFlags` was chosen back in the day since it described what `FEValues` should update when calling `reinit()` on a cell.

**Comment by bangerth** [comment]
On 1/25/22 11:48 AM, Daniel Arndt wrote:
> We can't easily error out for this case since the return value of |operator
> || is again |UpdateFlags| and I don't think we can change that either. In
> my opinion, the syntax in this particular case also makes sense since I
> would read
> 
> |flags.contains (update_values | update_gradients) |
> 
> as "flags contains update_values or update_gradients", i.e. at least one of
>  these flags. (Of course, this interpretation doesn't work with
> |operator&|).

That's precisely what I mean. If you can't express the |and| thing, then the 
function is poorly named. If you want to express

| flags & (update_values | update_gradients) |

then the correct naming of the function should probably be 
|.contains_any(...)| or |.contains_any_of(...)| instead of |.contains(...)|. 
The |any_of| spelling also matches an STL function.

What I meant to say above was that one can offer an abbreviated function 
|.contains(...)| that checks that only one bit is set, because then there is 
no ambiguity.

**Comment by masterleinad** [comment]
@bangerth I think it's rather the other way around. We can't prevent people from using
```
| flags.contains(update_values | update_gradients) |
```
but we can offer another function
```
| flags.contains_any_of(update_values, update_gradients) |
```
and then 
```
| flags.contains_all_of(update_values, update_gradients) |
```
Would your preference be to rather not define `contains` but only `contains_any_of` and `contains_all_of`?

**Comment by bangerth** [comment]
Yes, let's just not define `contains()` at all then. I am not sure whether we need `contains_all_of()` at all -- but you will know better, having looked at all of the places where it's used.

**Comment by tjhei** [comment]
What is the goal of this change? The poll in https://github.com/dealii/dealii/pull/13229#issuecomment-1014767396 seems to suggest that going back to the original state is currently in the majority.

Is the main advantage that we no longer have to use ``if (flags & (...))`` inside the library? The ``.contains(...)`` logic seems less clear to me (contains one of, contains any of?), while the bit logic is 100% clear to me without having to look up how something custom written works. 

Additionally, I find explaining a custom class with custom operator| and static const members more difficult to explain than "this is an enum with flags you can combine with |".

The only place ``if (flags & (...))`` is used is inside the library code, not in user code, so it all seems like a lot of work is being put into something that doesn't need improving (in my opinion).

**Comment by masterleinad** [comment]
@tjehi I would like to keep the clang-tidy check and was hoping this is a good compromise possibly with defining a bool conversion operator as described in the pull request description. It felt easier to proceed with the `contains` member function based on the other pull request.
If people are equally opposed to this pull request, I'm happy to close, of course.

**Comment by luca-heltai** [comment]
Personal opinion only here: as you might guess, I'm strongly against this change. This should remain an enum, or an enum class. 

One of the advantages of having an actual `enum` (or an `enum class`), is that we could leverage, when we get there, `magic_enum.hpp`.

My proposal was (along the line of) the following:

```
#include "parsed_tools/magic_enum.hpp"

template <typename T>
typename std::enable_if<std::is_enum<T>::value, bool>::type
operator && (T lhs, T rhs) // This could become: contains/contains_any_of/ operator , -- you name it 
{
  return (rhs & lhs) == lhs;
}


int
main()
{
  UpdateFlags flags = update_values | update_gradients;

  if (update_values && flags)
    std::cout << magic_enum::enum_name(update_values) << " is set. "
              << std::endl;
}
```

This would work also with our current implementation with no changes. We only need to make sure people understand that the operator `&&` is not commutative. If this is confusing, any function name here would work.

Let's not replicate `VtkFlags`. I think those are one of the worst thing we have in terms of maintainability, self documentation, and usage in the library. If you have ever written a parser for these types of objects, you know the pain.

**Comment by nfehn** [comment]
Taking into account previous comments and opinions on this topic, finding a good solution is difficult due to the fact that different developers mainly argue from on of the two perspectives (in my opinion):

- developers that care about the internal implementation: enum type should be avoided, implicit type conversions, boolean operations aren't a good programming style nowadays, bitwise operations are not intuitive to understand (code like `(update_flags & update_values) != 0u` is just cryptic in my opinion). Those developers prefer making changes to the code in order to improve the internal implementations of dealii, and might even accept incompatible changes
-  developers that care about backward compatibility and application codes, using `UpdateFlags` as e.g. in dealii tutorial step-3 (https://www.dealii.org/current/doxygen/deal.II/step_3.html): We have heard about the O(1000) occurrences of UpdateFlags in ASPECT, that @luca-heltai wants to parse expressions like `update_values | update_gradients | update_hessians`, etc.

Personally, I do not understand arguments by @tjhei like "The only place if (flags & (...)) is used is inside the library code" or @luca-heltai in https://github.com/dealii/dealii/issues/9888#issuecomment-613972406 "I don't really care how ugly the code looks like internally.".

I have the feeling that the argument "backward compatibility" kills the discussion. What I ask myself is why we do not try to combine the two perspectives described above. Here is my suggestion (a minimal solution):

```cpp
enum UpdateFlags
{
  update_default = 0,
  update_values = 0x0001,
  update_gradients = 0x0002,
  update_hessians = 0x0004
};

inline UpdateFlags
operator|(const UpdateFlags f1, const UpdateFlags f2)
{
  return static_cast<UpdateFlags>(static_cast<unsigned int>(f1) |
                                  static_cast<unsigned int>(f2));
}

class Flags
{
public:
  Flags()
  : values(false), gradients(false), hessians(false)
  {

  }

  // for backward compatibility reasons
  Flags(UpdateFlags const & update_flags)
  {
      values = update_flags & UpdateFlags::update_values;
      gradients = update_flags & UpdateFlags::update_gradients;
      hessians = update_flags & UpdateFlags::update_hessians;
  }

  bool values;
  bool gradients;
  bool hessians;
};

class FEValues
{
public:
  FEValues(Flags const & flags)
  {
    (void)flags;
  }
};

int
main(int argc, char ** argv)
{
  (void)argc;
  (void)argv;

  FEValues(UpdateFlags::update_values | UpdateFlags::update_gradients);

  return 0;
}
```

The two main benefits of this suggestion are
- that user code can remain unchanged (the ASPECT argument), and
- that dealii implementations become really readable and intuitive to understand, e.g.
```c
if (flags.values or flags.gradients or flags.hessians)
{
  ...
}
```
This suggestion therefore resolves the conflicting goals discussed previously in this PR and related PRs/issues.

**Comment by nfehn** [comment]
> What is the goal of this change? The poll in #13229 (comment) seems to suggest that going back to the original state is currently in the majority.

@tjhei Having read the above comment, I would have expected to find numbers larger than 2 and 4 in the poll that you mention. I took the freedom to also add my opinion to the poll. Regarding your question: I think several persons have given motivation for such a change ... Let me also add: As @drwells has written "If we determine some other workarounds we should update the poll.". In my opinion, other workarounds have been described.

**Comment by luca-heltai** [comment]
> Taking into account previous comments and opinions on this topic, finding a good solution is difficult due to the fact that different developers mainly argue from on of the two perspectives (in my opinion):
> 
> * developers that care about the internal implementation: enum type should be avoided, implicit type conversions, boolean operations aren't a good programming style nowadays, bitwise operations are not intuitive to understand (code like `(update_flags & update_values) != 0u` is just cryptic in my opinion). Those developers prefer making changes to the code in order to improve the internal implementations of dealii, and might even accept incompatible changes
> * developers that care about backward compatibility and application codes, using `UpdateFlags` as e.g. in dealii tutorial step-3 (https://www.dealii.org/current/doxygen/deal.II/step_3.html): We have heard about the O(1000) occurrences of UpdateFlags in ASPECT, that @luca-heltai wants to parse expressions like `update_values | update_gradients | update_hessians`, etc.

I think there is another essential point to add here, which is maintainability. We now got to the conclusion that we should avoid the pure enum type (even if I argue that this could be maintained for the most used flags, provided we eliminate the implicit bool conversion).

Maintaining backward compatibility and readability for our main flags (namely, the UpdateFlags) is something we should aim at as well. I personally think that breaking compatibility should be avoided at all costs, unless the new solution is, without discussion, a better design, or a *correct* design VS *buggy* design.

I still have codes that haunt me since we disallowed summing two point objects, and I agree with the change from the bottom of my heart. But it was a pain to convince my collaborators to switch to anything >= 8.5 precisely for this reason, and I don't like doing this to our users. 

The list of codes that I people should change for such an incompatibility is way too long to go with the policy "let's be brave and break compatibility" solution without thinking about it at least three times.

There is a possibility to break compatibility "only a little", for example, by converting to enum classes, introducing `operator &&`, and constexpr global variables. This is a solution I would prefer.

I don't think the alternative is maintainable. We have examples of that in the library:
```
VtKFlags flag;
falgs.write_high_order_cells;
data_out.set_flags(flags)
```
to me, this is very annoying. And I don't think that the following should be used
```
data_out.set_flags(VtKFlags(true, false, false,false,true));
```
because this does not document what I want to do, and if I read the program after a week, I have to look up what the `true, true, flase, false, true` mean, every single time. And if anyone decides to add a flag, or I decide to add a new flag, that is the last one in the constructor, this forces me to go the first route. And if flags are 4 or 5, all of a sudden a one liner (however ugly it can look like with enum classes) become a 10-liner. 

Extension of enum classes is trivial. Extension of a custom enum like class is not. You have to create a new member, a new constructor, a new parser option, etc. 

> 
> Personally, I do not understand arguments by @tjhei like "The only place if (flags & (...)) is used is inside the library code" or @luca-heltai in [#9888 (comment)](https://github.com/dealii/dealii/issues/9888#issuecomment-613972406) "I don't really care how ugly the code looks like internally.".
> 
> I have the feeling that the argument "backward compatibility" kills the discussion. 

Rightfully so, I would add. Breaking compatibility should be avoided, if possible. Period. Especially in features that *everybody* uses. When we changed `Point` it was a real pain. Now everyone knows how it works, but it took *years* to settle. If we changed `UpdateFalgs` in an incompatible way, this would have a terrible impact on our users. And on ourselves.

> What I ask myself is why we do not try to combine the two perspectives described above. Here is my suggestion (a minimal solution):
> 
> ```c++
> enum UpdateFlags
> {
>   update_default = 0,
>   update_values = 0x0001,
>   update_gradients = 0x0002,
>   update_hessians = 0x0004
> };
> 
> inline UpdateFlags
> operator|(const UpdateFlags f1, const UpdateFlags f2)
> {
>   return static_cast<UpdateFlags>(static_cast<unsigned int>(f1) |
>                                   static_cast<unsigned int>(f2));
> }
> 
> class Flags
> {
> public:
>   Flags()
>   : values(false), gradients(false), hessians(false)
>   {
> 
>   }
> 
>   // for backward compatibility reasons
>   Flags(UpdateFlags const & update_flags)
>   {
>       values = update_flags & UpdateFlags::update_values;
>       gradients = update_flags & UpdateFlags::update_gradients;
>       hessians = update_flags & UpdateFlags::update_hessians;
>   }
> 
>   bool values;
>   bool gradients;
>   bool hessians;
> };
> 
> class FEValues
> {
> public:
>   FEValues(Flags const & flags)
>   {
>     (void)flags;
>   }
> };
> 
> int
> main(int argc, char ** argv)
> {
>   (void)argc;
>   (void)argv;
> 
>   FEValues(UpdateFlags::update_values | UpdateFlags::update_gradients);
> 
>   return 0;
> }
> ```
> 
> The two main benefits of this suggestion are
> 
> * that user code can remain unchanged (the ASPECT argument), and
> * that dealii implementations become really readable and intuitive to understand, e.g.
> 
> ```c
> if (flags.values or flags.gradients or flags.hessians)
> {
>   ...
> }
> ```
> 
> This suggestion therefore resolves the conflicting goals discussed previously in this PR and related PRs/issues.

I think this all started because people found ugly the expression (as you pointed out):
```
(flags & (update_values | update_gradients)) != 0u;
```
I argue that we can avoid these explicit conversions by an `operator &&` *without any further changes*, 
```
flags && (update_values | update_gradients)
```
which would keep `UpdateFlags` as an enum. I am also not against making it an enum class, with some backward compatible definitions to help user codes.

However, having to maintain two different classes (an enum class, and a Flags class) is not, in my absolutely personal opinion, a maintainable solution, **unless** you can come up with a solution where 

- the additional class is generated automatically from the flags (via some template magic/preprocessing magic, similarly to what is done in magic_enum.hpp)
- the additional class must be a standard one, for all enums, and could be used as 
```
template<typename T>
class BitWiseHelper {
 BitWiseHelper() {
static_assert(std::is_enum<T>::value, "Not implemented for non-enum types);
// automatically fill members
// maybe define here the bitwise operators.
}

using UpdateFlags = BitWiseHelper<internal::UpdateFlags>;
```
Only in this case I would agree that we are putting together the two design concepts, and getting the best out of it. Anything that needs to be implemented by hand *for each enum class* I would avoid at all costs.

**Comment by tamiko** [comment]
The Ci is failing with:
```
/jenkins/workspace/dealii_PR-13290/include/deal.II/matrix_free/mapping_data_on_the_fly.h:215:11: error: could not convert ‘((dealii::internal::MatrixFreeFunctions::MappingDataOnTheFly<2, dealii::VectorizedArray<double, 2ul> >*)this)->dealii::internal::MatrixFreeFunctions::MappingDataOnTheFly<2, dealii::VectorizedArray<double, 2ul> >::fe_values.dealii::FEValues<2, 2>::<anonymous>.dealii::FEValuesBase<dim, spacedim>::get_update_flags<2, 2>().dealii::UpdateFlags::operator&(dealii::update_JxW_values)’ from ‘dealii::UpdateFlags’ to ‘bool’

           if (fe_values.get_update_flags() & update_JxW_values)

           ^

```

**Comment by tamiko** [comment]
Dear all, I have locked the conversation on this pull request for 24h. I would also like to ask all developers to not comment for a while until the lock is lifted for everyone.

I would like to request everyone to calm down.
In particular, it is not necessary to attack individuals with baseless insinuations, or questioning their motives.
This is especially true for this discussion about a minor technical detail that boils down to taste.

**Comment by tamiko** [comment]
Dear all, I have unlocked the conversation again.

**Comment by nfehn** [comment]
One of the suggestions included the following function:

>
```c
template <typename T>
typename std::enable_if<std::is_enum<T>::value, bool>::type
operator && (T lhs, T rhs) // This could become: contains/contains_any_of/ operator , -- you name it 
{
  return ((rhs & lhs) == lhs) ||  ((rhs & lhs) == rhs));
}
```
I would like to ask how the documentation of this function would be formulated in written text?

**Comment by masterleinad** [comment]
> I argue that we can avoid these explicit conversions by an `operator &&` _without any further changes_,
> 
> ```
> flags && (update_values | update_gradients)
> ```
> 
> which would keep `UpdateFlags` as an enum. I am also not against making it an enum class, with some backward compatible definitions to help user codes.

I would be happy with that, too.

**Comment by bangerth** [comment]
I am opposed to overloading `operator &&`. I think that does not address the readability issue at all since it introduces the operator in a meaning that is even less widely understood than the bit-twiddle `operator &`.

**Comment by masterleinad** [comment]
> I am opposed to overloading `operator &&`. I think that does not address the readability issue at all since it introduces the operator in a meaning that is even less widely understood than the bit-twiddle `operator &`.

It's certainly not my favorite either but it would be a compromise to me to keep the `clang-tidy` check.

**Comment by luca-heltai** [comment]
> ```c
> operator && (T lhs, T rhs)
> ```
> 
> I would like to ask how the documentation of this function would be formulated in written text?

I would formulate the documentation similarly to: 

> Checks if the two bitwise flags passed as argument are one a superset of the other. This is equivalent to explicitly converting to `bool` the result of `lhs & rhs`, i.e.,  `static_cast<bool>(lhs & rhs)`

**Comment by kronbichler** [comment]
>  This is equivalent to explicitly converting to `bool` the result of `lhs & rhs`, i.e., `static_cast<bool>(lhs & rhs)`

I have not followed all of the discussion and am trying to catch up only, but is this true for the implementation we were looking at? In the `static_cast?` case, I get true when both `lhs` and `rhs` have one bit in common. Think of `lhs = update_values | update_gradients` and `rhs = update_values | update_JxW_values`, for example. However, when I do
```c++
template <typename T>
typename std::enable_if<std::is_enum<T>::value, bool>::type
operator && (T lhs, T rhs) // This could become: contains/contains_any_of/ operator , -- you name it 
{
  return ((rhs & lhs) == lhs) ||  ((rhs & lhs) == rhs));
}
```
Then I get false, because `rhs & lhs` becomes `update_values` in a bitwise and operation, and is hence not equal to any of `lhs` or `rhs`, and hence we would return `false`. I assume you really want `operator &&` (or whatever name we want to have) to return `true` in this case as we want to create a `contains_any_of` operation, so it is just the code that is presented here that needs a somewhat different implementation?

Either way, I will be fine with whatever decision will come out of this discussion (as long as I can vote somewhere for my preference), and it's good to have all arguments on the table when making such a decision.

**Comment by nfehn** [comment]
I agree with @bangerth that the `&&` operator should not be used for what we need here. This function needs to be documented well in order to not be misused, so I see a clear value in using a function name that tells us what it does in words. Regarding function names, I think I would prefer something like `have_intersection(a,b)` (in case of a free function), or `a.has_intersection_with(b)` in case of a member function.

**Comment by luca-heltai** [comment]
> > This is equivalent to explicitly converting to `bool` the result of `lhs & rhs`, i.e., `static_cast<bool>(lhs & rhs)`

Indeed you are right. This part of the documentation is wrong. That algorithm implements the concept of superset of bits.

> Then I get false, because `rhs & lhs` becomes `update_values` in a bitwise and operation, and is hence not equal to any of `lhs` or `rhs`, and hence we would return `false`. I assume you really want `operator &&` (or whatever name we want to have) to return `true` in this case as we want to create a `contains_any_of` operation, so it is just the code that is presented here that needs a somewhat different implementation?

I think what we need in the code is possibility to check if one of the flag is entirely contained in the other flag, independently on the order in which they are presented. With a function like `contains_all`, we'd probably want to have to specify the order correctly, i.e., `contains_all(left, right)` would return in general a different value than `contains_all(right, left)`.

**Comment by luca-heltai** [comment]
Also `have_intersection` or `has_intersection` would be fine for me. I liked the `&&` only because the outcome of an `operator &&` is a bool, but I don't really care if we give it a more descriptive name.

**Comment by tjhei** [comment]
A very strong "no" from me to replace the bit-wise & by ``operator &&``. This is non-standard, a lot less intuitive, and impossible to learn about when reading the code. 
Imagine a bool ``need_recalculation`` and a statement like ``if (need_recalculation && update_flag && update_values)``...

**Comment by nfehn** [comment]
While I would not use `&&` as function name, I want to point out that the suggestion by @luca-heltai still seems to be the one that allows to solve the problem at hand and takes care of the different interests (backward compatibility, clang-tidy checks, does not have the design problems of the enum-like class described above https://github.com/dealii/dealii/pull/13290#issuecomment-1021311981)
>The following aspects might be considered a disadvantage of the proposed design:
>
> - A private constructor and a function declared "friend" is needed
>
> - For a variable UpdateFlags flags;, one could write flags.update_values.update_gradients (potentially continuing this chain further), since sub-elements are again of the same type UpdateFlags (flags are nested), or do I miss something?

In this sense, the poll in https://github.com/dealii/dealii/pull/13229#issuecomment-1014767396 (the last option named "operator `&&`") should not be taken literally. The variant "for the previous status quo: `if (flags & update_hessians)` and disabling implicit bool checks" has a relative majority with 8 supporters currently, but the absolute majority does not seem so clear.

Therefore, I would suggest another procedure. We should first make a poll whether we want to have the clang-tidy checks or not. Next, we discuss how to realize this. Backward-compatible suggestions are on the table.

**Comment by drwells** [comment]
Certain options (like 'return to status quo') already imply whether or not we can keep the implicit boolean conversions or not. `bool operator&&(UpdateFlags a, UpdateFlags b)` and `bool is_subset(UpdateFlags a, UpdateFlags b)` resolve the problem in a similar way so I combined those two options in the existing poll (and that would allow us to enable implicit boolean conversion checks) - if that one wins we can vote some more on which one to pick :)

I would prefer not to add a new poll after so many people have voted in the open one. In particular, at this point, it really does look like the previous status quo is the winner by a large margin even though that means missing out on some clang checks.

**Comment by nfehn** [comment]
Let me first point out that I am fine with whatever decision will come out (I made my personal point of view clear elsewhere). However, I also want to mention that my impression from the discussions in three issues/PRs (https://github.com/dealii/dealii/pull/13150#discussion_r778209647, https://github.com/dealii/dealii/pull/13229, https://github.com/dealii/dealii/pull/13290) is that the majority of developers really discusses about how to improve the library for the good. @drwells, regarding your comment (https://github.com/dealii/dealii/pull/13229#issuecomment-1023243872):

> I think enough people are in favor of returning to "use enums and disable extra clang-tidy checks" that we should make that decision now.

When browsing again through the three issues, I mainly found these two comments supporting this point: https://github.com/dealii/dealii/pull/13229#issuecomment-1012197104 and https://github.com/dealii/dealii/pull/13290#issuecomment-1021661504 (I think I have seen more comments that say that the use of ``&`` is not intuitive to understannd). The other comments - in my opinion - essentially discuss solutions to the problem raised/addressed by @masterleinad. I consider the "backward compatibility" argument somewhat ruled out as enough suggestions have been made up to now that resolve this. In summary, my concern is just that a solution is "elected" because it is "most widely known", not because it is the best.

**Comment by nfehn** [comment]
As a kind of "late feedback" on the topics discussed here (it typically needs some time that things get adjusted in user code ...):

In comment https://github.com/dealii/dealii/pull/13290#issuecomment-1021661504, it was stated that the bitwise logic only affects the internal implementation in deal.II, not user code. It was one of the arguments to revert back to the old solution with & and | operators. In this PR https://github.com/exadg/exadg/pull/243 (where we have to change user code because of a deprecation warning, which can be seen as an incompatibility in practice in my opinion), however, we observe that the discussions here essentially translate to user code.

**Pull request closed without merging**